pommerman/forward_model.py

```diff
@@ -156,128 +156,256 @@ def step(actions,
-                    if not utility.position_is_bomb(curr_bombs, position):
-                        bomb = agent.maybe_lay_bomb()
-                        if bomb:
-                            curr_bombs.append(bomb)
-                elif utility.is_valid_direction(curr_board, position, action):
-                    next_position = agent.get_next_position(action)
-
-                    # This might be a bomb position. Only move in that case if
-                    # the agent can kick.
-                    if not utility.position_is_bomb(curr_bombs, next_position):
-                        next_positions[agent.agent_id] = next_position
-                    elif not agent.can_kick:
-                        agent.stop()
-                    else:
-                        after_next_position = utility.get_next_position(
-                            next_position, constants.Action(action))
-                        if not utility.position_on_board(
-                                curr_board, after_next_position
-                        ) or not utility.position_is_passage(
-                                curr_board, after_next_position):
-                            agent.stop()
-                        else:
-                            next_positions[agent.agent_id] = next_position
+        for bomb_num, bomb in enumerate(curr_bombs):
+            curr_board[bomb.position] = constants.Item.Passage.value
+            if bomb.is_moving():
+                desired_position = utility.get_next_position(
+                    bomb.position, bomb.moving_direction)
+                if utility.position_on_board(
+                        curr_board, desired_position) and not (
+                            curr_board[desired_position] in items_and_walls):
+                    desired_bomb_positions[bomb_num] = desired_position
+
+        # Deal with position switches:
+        # Agent <-> Agent => revert to previous position
+        # Bomb <-> Bomb => revert to previous position
+        # Agent <-> Bomb => revert Bomb only to previous position (for nice kicks)
+
+        crossings = dict()
+
+        def crossing(current, desired):
+            current_x, current_y = current
+            desired_x, desired_y = desired
+            if current_x != desired_x:
+                assert current_y == desired_y
+                return ('X', min(current_x, desired_x), current_y)
+            assert current_x == desired_x
+            return ('Y', current_x, min(current_y, desired_y))
+
+        for agent_num, agent in enumerate(alive_agents):
+            if desired_agent_positions[agent_num] != agent.position:
+                border = crossing(agent.position,
+                                  desired_agent_positions[agent_num])
+                if border in crossings:
+                    desired_agent_positions[agent_num] = agent.position
+                    num, isAgent = crossings[border]
+                    assert isAgent
```

Comment:
when will this be triggered?

Reply:
This prevents agents and bombs switching positions.
The idea here is a collision map for the borders between two squares.
If the border is crossed twice two objects (agents or bombs) try to switch positions

Reply:
I understand that. I'm asking when would isAgent ever be False?

Reply:
In theory never - in practice it prevents strange things from happening if code ever gets changed and agent and bomb iterations switch positions.